### PR TITLE
Restrict access for different users roles

### PR DESCRIPTION
**Pull request by alghazu**
### What we did:
- Add acceptance test
- Create users roles using enum
- Add Pundit gem and generate policy for articles
- Add approval attribute to articles and comments
- Fix relation issues in models
- Restrict users access as per their roles

### CODE REVIEW

File: features/restrict_different_users_roles.feature [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  In order to restrict how much access different users have
+  I would like to add different user roles
+
+Background:
+  Given the following users exist
+    | email                 | role        |
+    | editor@test.com       | editor      |
+    | journalist@test.com   | journalist  |
+    | subscriber@test.com   | subscriber  |
+  Given we have the following categories
```
Comment (tochman) [inline features/restrict_different_users_roles.feature]: Insert line break

File: features/restrict_different_users_roles.feature [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  Given I am signed in as "subscriber@test.com"
+  And I am on the landing page
+  Then I should not see "Create Article"
+
+Scenario: Subscriber can NOT create articles by accessing New Article path
+  Given I am signed in as "subscriber@test.com"
+  When I visit the new article page
+  Then I should see "You are not authorized to perform this action!"
+
+Scenario: Journalist creates an article
```
Comment (tochman) [inline features/restrict_different_users_roles.feature]: "Journalist can create an article"

File: features/restrict_different_users_roles.feature [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  Given I am signed in as "journalist@test.com"
+  And I am on the landing page
+  When I click "Create Article"
+  When I fill in "Headline" with "The new article"
+  And I fill in "Content" with "This is the new text content"
+  And I select "Sports" from categories menu
+  And I click "Save Article"
+  Then I should see "Article successfully saved and sent for approval"
+  When I am on the landing page
```
Comment (tochman) [inline features/restrict_different_users_roles.feature]:
```'
But when I am on the landing page'
```

File: features/restrict_different_users_roles.feature [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  Then I should not see "The new article"
+
+Scenario: Editor creates an article then deletes it
```
Comment (tochman) [inline features/restrict_different_users_roles.feature]: Please, leave the destroy part of this scenario out

File: features/user_can_edit_articles.feature [hunk 5]
```diff
@@ -4,10 +4,15 @@ Feature: Journalist can edit articles
...
-  Given we have the following articles
-    | headline                  | content     |
-    | The awesome article       | Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. |
-  And user is signed in
+  Given the following users exist
+    | email                 | role        |
+    | journalist@test.com   | journalist  |
+
+  And we have the following articles
+    | headline            | content                    | user                 | approval |
```
Comment (tochman) [inline features/user_can_edit_articles.feature]: Revisit the choice of attribute name: `approval`

File: app/policies/article_policy.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    user.editor? || user.journalist?
+  end
+
+  def update?
+    user.editor? || (user.journalist? && record.user == user)
+  end
+
+  def destroy?
+    user.editor?
+  end
```
Comment (tochman) [inline app/policies/article_policy.rb:21]: Please test with acceptance tests.

File: app/policies/application_policy.rb [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    @user = user
+    @record = record
+  end
+
+  def index?
+    false
+  end
+
+  def show?
+    scope.where(:id => record.id).exists?
```
Comment (tochman) [inline app/policies/application_policy.rb]: Scaffolded code often uses old hash syntax. Stay frosty!

**Review by tochman — CHANGES_REQUESTED** [review]
Please see comments

**Comment by alghazu** [comment]
@tochman comments addressed 
In addition, four feature files created:
- users_permissions_for_article_creation
- users_permissions_for_article_edit
- users_permissions_for_article_delete
- users_permissions_for_comment_creation

File: app/policies/application_policy.rb [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    def initialize(user, scope)
+      @user = user
+      @scope = scope
+    end
+
+    def resolve
+      scope
+    end
+  end
+end
```
Comment (tochman) [inline app/policies/application_policy.rb]: Do you really need the ApplicationPolicy to hold so much code?

File: app/views/layouts/_navbar.html.haml [hunk 9]
```diff
@@ -1,3 +1,8 @@
-= link_to "Create Article", new_article_path
-= link_to "Sign In", new_user_session_path
-= link_to "Sign Up", new_user_registration_path
+- if user_signed_in?
+  = "Signed in as: #{current_user.role.capitalize} (#{current_user.email}) |"
+  = link_to("Create Article", new_article_path) + "|" if policy(Article).new?
```
Comment (tochman) [inline app/views/layouts/_navbar.html.haml]: Please be consistent with the usage of parathesis

File: features/article_are_grouped_in_categories.feature [hunk 10]
```diff
@@ -10,9 +10,9 @@ Background:
     | Technology    |
 
   And we have the following articles
-    | headline                            | category   |
-    | The sport team won something        | Sports     |
-    | Some programers made some new stuff | Technology |
+    | headline                            | category   | published |
+    | The sport team won something        | Sports     | true     |
+    | Some programers made some new stuff | Technology | true     |
```
Comment (tochman) [inline features/article_are_grouped_in_categories.feature]: You can let the Factory deal with the `published` attribute if you are not using it in the scenario. Would lead to less code == less changed files == less work for you == less work for me ;-)

File: features/user_can_edit_articles.feature [hunk 11]
```diff
@@ -4,10 +4,15 @@ Feature: Journalist can edit articles
   I would like to be able to edit and resubmit my articles
 
 Background:
-  Given we have the following articles
-    | headline                  | content     |
-    | The awesome article       | Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. |
-  And user is signed in
+  Given the following users exist
+    | email                 | role        |
+    | journalist@test.com   | journalist  |
```
Comment (tochman) [inline features/user_can_edit_articles.feature:9]: If the mail character is a Journalist, then please rename this file to reflect that

File: features/users_permissions_for_article_creation.feature [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  When I click "Create Article"
+  When I fill in "Headline" with "The new article"
+  And I fill in "Content" with "This is the new text content"
+  And I select "Sports" from categories menu
+  And I click "Save Article"
+  Then I should see "Article was successfully saved and sent for approval"
+  But when I am on the landing page
+  Then I should not see "The new article"
+
+Scenario: Editor can creates an article
```
Comment (tochman) [inline features/users_permissions_for_article_creation.feature]: creates -> create

File: spec/factories/users.rb [hunk 13]
```diff
@@ -1,6 +1,7 @@
 FactoryBot.define do
-  factory :user do
-    email "test@test.com"
-    password "1234567"
+  factory :user do |u|
+    u.sequence(:email) { |n| "journalist#{n}@test.com"}
+    u.password "password123"
+    u.role 'journalist'
```
Comment (tochman) [inline spec/factories/users.rb]: Is the journalist the most common role. Also, why the change in notation?

File: app/views/layouts/_navbar.html.haml (hunk 9 above)
Comment (alghazu) [reply app/views/layouts/_navbar.html.haml]: We skip the parenthesis when possible to take advantage of the beauty of ruby.
But in this specific case, they are required so the pipe " | " gets involved in the conditional statement.

File: features/article_are_grouped_in_categories.feature (hunk 10 above)
Comment (alghazu) [reply features/article_are_grouped_in_categories.feature]: Done, looks much cleaner now.

File: spec/factories/users.rb (hunk 13 above)
Comment (alghazu) [reply spec/factories/users.rb]: changed to `subscriber` and modified one legacy test.

File: app/policies/application_policy.rb (hunk 8 above)
Comment (tochman) [reply app/policies/application_policy.rb]: Well, do you?

File: app/models/user.rb [hunk 14]
```diff
@@ -1,6 +1,18 @@
...
          :recoverable, :rememberable, :trackable, :validatable
+
+  validates :role, presence: true
+  has_many :articles
+  has_many :comments
+
+  private
+
+    def set_default_role
+      self.role ||= :subscriber
```
Comment (tochman) [inline app/models/user.rb:16]: Are you a subscriber as soon as you sign up?

File: app/policies/application_policy.rb (hunk 8 above)
Comment (alghazu) [reply app/policies/application_policy.rb]: We deleted the unused methods

File: app/models/user.rb (hunk 14 above)
Comment (ghost) [reply app/models/user.rb:16]: When someone Sign Up, by default is a subscriber. 
To get a Journalist or Editor account, needs to contact us.

Comment (tochman) [reply app/models/user.rb:16]: I thought that you become a subscriber if you pay for a subscription, but that is not an issue on this PR. Not a blocker.

File: app/policies/application_policy.rb (hunk 8 above)
Comment (tochman) [reply app/policies/application_policy.rb]: I know that you can cut away more on this. Not a blocker though (unless your test coverage drops)

**Review by tochman — APPROVED** [review]

File: app/policies/application_policy.rb (hunk 8 above)
Comment (alghazu) [reply app/policies/application_policy.rb]: removed all unused parts of Pundit scaffold
